src/cache/src/cache.c

```diff
@@ -0,0 +1,297 @@
+static int cacheInitItemMutex(cache_t*);
+static void cacheFreeChunkItems(cache_t*);
+static void cacheFreeMemory(cache_t*);
+static void cacheFreeTable(cache_t*);
+static void cacheDestroyItemMutex(cache_t*);
+
+cache_t* cacheCreate(cacheOption* options) {
+  if (check_cache_options(options) != CACHE_OK) {
+    cacheError("check_cache_options fail");
+    return NULL;
+  }
+
+  cache_t* pCache = calloc(1, sizeof(cache_t));
+  if (pCache == NULL) {
+    cacheError("calloc cache_t fail");
+    return NULL;
+  }
+
+  taosInitRWLatch(&(pCache->latch));
+  pCache->options = *options;
+
+  if (cacheSlabInit(pCache) != CACHE_OK) {
+    goto error;
+  }
+
+  if (cacheLruInit(pCache) != CACHE_OK) {
+    goto error;
+  }
+
+  if (cacheInitItemMutex(pCache) != CACHE_OK) {
+    goto error;
+  }
+
+  return pCache;
+
+error:
+  if (pCache != NULL) {
+    free(pCache);
+  }
+
+  return NULL;
+}
+
+void  cacheDestroy(cache_t* pCache) {
+  cacheSlabDestroy(pCache);
+  cacheLruDestroy(pCache);
+  cacheFreeChunkItems(pCache);
+  cacheFreeMemory(pCache);
+  cacheFreeTable(pCache);
+  cacheDestroyItemMutex(pCache);
+
+  free(pCache);
+}
+
+int cachePut(cacheTable* pTable, const void* key, uint8_t nkey, const void* value, uint32_t nbytes, uint64_t expire) {
+  cacheMutex* pMutex = getItemMutexByKey(pTable, key, nkey);
+  cacheMutexLock(pMutex);
+
+  int ret = cachePutDataIntoCache(pTable,key,nkey,value,nbytes, pMutex, NULL, expire);
+
+  cacheMutexUnlock(pMutex);
+
+  return ret;
+}
+
+void* cacheGet(cacheTable* pTable, const void* key, uint8_t nkey, int* nbytes) {
+  cacheMutex *pMutex = getItemMutexByKey(pTable, key, nkey);
+  cacheMutexLock(pMutex);
+
+  if (nbytes) {
+    *nbytes = 0;
+  }
+  /* first find the key in the cache table */
+  cacheItem* pItem = cacheTableGet(pTable, key, nkey);
+  if (pItem) {
+    itemIncrRef(pItem);
+    uint64_t now = taosGetTimestamp(TSDB_TIME_PRECISION_MILLI);
+    #if 0
+    if (cacheItemIsExpired(pItem, now)) { /* is item expired? */
+      /* cacheItemUnlink make ref == 1 */
+      cacheItemUnlink(pTable, pItem, CACHE_LOCK_LRU);
+      /* cacheItemRemove make ref == 0 then free item */
+      cacheItemRemove(pTable->pCache, pItem);
+      pItem = NULL;
+    } else
+    #endif
+    if (cacheItemIsNeverExpired(pItem)) {      
+      /* never expired item refCount == 1 */
+      assert(itemRef(pItem) == 2);
+      pItem->lastTime = now;    
+    } else {
+      cacheItemBump(pTable, pItem, now);      
+    }
+    goto out;
+  }
+
+  /* try to load the data from user defined function */
+  if (pTable->option.loadFp == NULL) {
+    goto out;
+  }
+  char *loadValue;
+  size_t loadLen = 0;
+  uint64_t expire;
+  if (pTable->option.loadFp(pTable->option.userData, key, nkey, &loadValue, &loadLen, &expire) != CACHE_OK) {
+    goto out;
+  }
+
+  /* TODO: save in the cache if access only one time? */
+  int ret = cachePutDataIntoCache(pTable,key,nkey,loadValue,loadLen,pMutex, &pItem, expire);
+  free(loadValue);
```

Comment:
Since loadValue is passed into a user defined function, it's better to init loadValue with NULL and use tfree here.

Reply:
Or maybe add a deleteFP provided by user?